core/threaded/scheduler_GEDF_NP.c

```diff
@@ -223,91 +105,138 @@ void lf_sched_init(environment_t* env, size_t number_of_workers, sched_params_t*
+  pqueue_free((pqueue_t*)scheduler->custom_data->reaction_q);
+  free(scheduler->custom_data);
 }
 
 ///////////////////// Scheduler Worker API (public) /////////////////////////
-/**
- * @brief Ask the scheduler for one more reaction.
- *
- * This function blocks until it can return a ready reaction for worker thread
- * 'worker_number' or it is time for the worker thread to stop and exit (where a
- * NULL value would be returned).
- *
- * @param worker_number
- * @return reaction_t* A reaction for the worker to execute. NULL if the calling
- * worker thread should exit.
- */
+
 reaction_t* lf_sched_get_ready_reaction(lf_scheduler_t* scheduler, int worker_number) {
-  // Iterate until the stop_tag is reached or reaction queue is empty
-  while (!scheduler->should_stop) {
-    // Need to lock the mutex for the current level
-    size_t current_level = scheduler->next_reaction_level - 1;
-    LF_PRINT_DEBUG("Scheduler: Worker %d trying to lock the mutex for level %zu.", worker_number, current_level);
-    LF_MUTEX_LOCK(&scheduler->array_of_mutexes[current_level]);
-    LF_PRINT_DEBUG("Scheduler: Worker %d locked the mutex for level %zu.", worker_number, current_level);
-    reaction_t* reaction_to_return = (reaction_t*)pqueue_pop((pqueue_t*)scheduler->executing_reactions);
-    LF_MUTEX_UNLOCK(&scheduler->array_of_mutexes[current_level]);
+  // Need to lock the environment mutex.
+  LF_PRINT_DEBUG("Scheduler: Worker %d locking environment mutex.", worker_number);
+  LF_MUTEX_LOCK(&scheduler->env->mutex);
+  LF_PRINT_DEBUG("Scheduler: Worker %d locked environment mutex.", worker_number);
 
+  // Iterate until the stop_tag is reached or the event queue is empty.
+  while (!scheduler->should_stop) {
+    reaction_t* reaction_to_return = (reaction_t*)pqueue_peek(scheduler->custom_data->reaction_q);
     if (reaction_to_return != NULL) {
-      // Got a reaction
-      return reaction_to_return;
+      // Found a reaction.  Check the level.  Notice that because of deadlines, the current level
+      // may advance to the maximum and then back down to 0.
+      if (LF_LEVEL(reaction_to_return->index) == scheduler->custom_data->current_level) {
+        // Found a reaction at the current level.
+        LF_PRINT_DEBUG("Scheduler: Worker %d found a reaction at level %zu.", worker_number,
+                       scheduler->custom_data->current_level);
+        // Remove the reaction from the queue.
+        pqueue_pop(scheduler->custom_data->reaction_q);
+
+        // If there is another reaction at the current level and an idle thread, then
+        // notify an idle thread.
+        reaction_t* next_reaction = (reaction_t*)pqueue_peek(scheduler->custom_data->reaction_q);
+        if (next_reaction != NULL && LF_LEVEL(next_reaction->index) == scheduler->custom_data->current_level &&
+            scheduler->number_of_idle_workers > 0) {
+          // Notify an idle thread.
+          LF_COND_SIGNAL(&scheduler->custom_data->reaction_q_changed);
+        }
```

Comment:
I do not understand why an individual other worker is being signaled to here, after we are already processing a level. Why doesn't the worker that takes charge of advancing the level broadcast to all other workers then? Then they would all be woken up front, in a relatively cold path, instead of doing it piecemeal, in a relatively hot path.